File: Rust/p004.rs
/*
A palindromic number reads the same both ways. The largest palindrome made
from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

fn main() {
    let answer = brute_force();
    println!("{answer}");
}

fn brute_force() -> u32 {
    let mut max = 0;
    let mut prod;

    for x in 100..1000 {
        for y in 100..1000 {
            prod = x * y;
            if prod > max && is_palindrome(prod) {
                max = prod;
            }
        }
    }

    max
}

fn is_palindrome(n: u32) -> bool {
    let digits: Vec<u8> = n
        .to_string()
        .chars()
        .map(|ch| ch.to_digit(10).unwrap() as u8)
        .collect();

    let len = digits.len();
    let half_len = len / 2;

    for i in 0..half_len {
        if digits[i] != digits[len - 1 - i] {
            return false;
        }
    }

    true
}

File: Python/p004.py
"""
A palindromic number reads the same both ways. The largest palindrome made
from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
"""

from utils import is_palindrome


def brute_force():
    iterator = (x * y for y in range(100, 999+1) for x in range(100, 999+1))
    return max(n for n in iterator if is_palindrome(n))


if __name__ == '__main__':
    answer = brute_force()
    print(answer)

